### PR TITLE
Array slices

### PR DESCRIPTION
**Pull request by glyn**

### CODE REVIEW

**Comment by gregsdennis** [comment]
I think we need to add an "optional" property to the tests (or break them out to another section).  The "big num" cases can't be supported by all frameworks, and we shouldn't expect it of them.  There _will_ be other cars like this as well.

File: src/ast.rs
Comment (mkmik) [inline src/ast.rs:128]: I tried
```rust
array_slice(j.as_array().unwrap(), 0, -1, 1);
```

and function never terminates (see https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8f0f37bffb80a1e5d4bd23137cfc7d3d)

Comment (mkmik) [reply src/ast.rs:128]: it's not clear to me from the spec whether a negative "end" is meant to be interpreted as `len + end` or if it just means that `i < end` is always false and thus the slice is always empty

File: src/ast.rs
Comment (mkmik) [inline src/ast.rs:141]: The spec says:

```
A negative index j selects an element of an array of length len if and only if 0 <= j + len < len, in which case it selects the same element as the non-negative index j + len.
```

then it describes the iterations as:

```
for (i = start; i < end; i = i + step) {
  ...
}
```

what should happen if you call this on an input slice of length 5?:
```rust
    let s = array_slice(j.as_array().unwrap(), -5, 0, 1);
```

My interpretation is that it would yield

```
i = -5;
i < 0 //  true
push(arr[i + len]) // arr[-5+5] === arr[0]
i = i+1 //  -4
i < 0 //  true
push(arr[i + len]) // arr[-4+5] === arr[1]
...
```

**Comment by glyn** [comment]
> I think we need to add an "optional" property to the tests (or break them out to another section). The "big num" cases can't be supported by all frameworks, and we shouldn't expect it of them. There _will_ be other cars like this as well.

The spec is currently silent about supported precisions of array indices. Probably needs firming up in some way.

It's interesting because for any given language or machine architecture, there will exist large values which exceed the built-in capabilities of the language or architecture.

> 
> I know we have a "skip" setting, but that means modifying the CTS after I've downloaded it. That should be considered bad practice. The suite should run as-is.

Depending on how the spec pans out, one option would be to add some kind of "category" field to the tests so that those which may be optional in some sense can be skipped _en masse_. I'm reticent to do that prematurely. Why not clamp the values to a suitable range (e.g. signed 31 bit) for now? Or is that too inconvenient?

File: src/ast.rs
Comment (glyn) [reply src/ast.rs:128]: > I tried
> 
> ```rust
> array_slice(j.as_array().unwrap(), 0, -1, 1);
> ```
> 
> and function never terminates (see https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8f0f37bffb80a1e5d4bd23137cfc7d3d)

Who needs fuzz testing with reviewers like @mkmik? Great catch!

File: src/ast.rs
Comment (glyn) [reply src/ast.rs:141]: This is a good example of a reference implementation and compliance test suite turning up a hole in the spec.

I think the spec _should_ say (only more formally/precisely) that negative `start` is syntactic sugar for `start + len` and similarly for negative `end`. But then `start` and `end` need to be desugared _before_ plugging them into the relevant `for` loop.

So, for an array of length 5, `[-5, 0, 1]` corresponds to the elements indexed by the values of `i` in the `for` loop:
```
for (i = 0; i < 0; i = i + 1) {
  ...
}
```
of which there are none, so the result is empty.

Comment (mkmik) [reply src/ast.rs:141]: Desugaring before iterating makes most sense to me as well.

A way tinkering with the consequences if doing it the other way around and I don't think it would provide any advantage. An interesting consequence is that it allows the resulting slice to be longer than the input array (and contain repeated elements) which I think would be quite confusing.

File: src/ast.rs
Comment (glyn) [reply src/ast.rs:128]: Actually, I can't drive this case externally because the caller of `array_slice` handles negative `start` and `end` values. So `array_slice` has some unstated preconditions.

Comment (glyn) [reply src/ast.rs]: @mkmik I pushed that logic inside `array_slice` to make it total.

File: src/ast.rs
Comment (glyn) [reply src/ast.rs:143]: I think the implementation is correct and we just need to fix the spec. ;-)

Comment (mkmik) [reply src/ast.rs:143]: the privileges of a spec writer! :-)

Comment (mkmik) [reply src/ast.rs:143]: I wonder if something like this would have less special casing: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dd7180d6586e5a511b4e56b0058e5562

**Comment by gregsdennis** [comment]
I'm happy for this to merge for now. We can come back to it when the spec addresses it.  Let's discuss in your new issue.

File: src/ast.rs
Comment (glyn) [reply src/ast.rs]: This is all moot after rebasing on slyce.

**Comment by glyn** [comment]
> I'm happy for this to merge for now. We can come back to it when the spec addresses it. Let's discuss in your new issue.

I agree the code is good, but I'd prefer to maintain consistency with the spec. I'll merge this PR once the spec is updated.

File: src/parser.rs
Comment (mkmik) [inline src/parser.rs:100]: Ah, I see now why you wanted to add the `Default` trait impl into slyce. It was not strictly necessary (although it doesn't hurt), since the `From` trait implementations also come with a blanket impl of `Into`:

```suggestion
        start: start.into(),
        end: end.into(),
```

the slyce Readme shows how `into()` can be used, but it's a bit artificial in one-liners because for type
inference to work seamlessly you need variables with real types that force the hand (like in this case `start`, ...)

**Comment by glyn** [comment]
Blocked on https://github.com/jsonpath-standard/internet-draft/pull/31

**Comment by glyn** [comment]
@gregsdennis I'd appreciate your review of the final commit which tests the behaviour when integer representations of array indices and slice parameters overflow. This aims to enforce the current spec.

**Review by gregsdennis — COMMENTED** [review]
Shouldn't the implementation at least attempt to parse the invalid selectors?  It looks like cts.rs is just skipping ones that are marked as invalid.

**Comment by glyn** [comment]
@gregsdennis wrote:

> Shouldn't the implementation at least attempt to parse the invalid selectors? It looks like cts.rs is just skipping ones that are marked as invalid.

I'm not sure what gives that impression. The `parse` function of the implementation is executed in the code below from cts.rs regardless of whether the testcase says the selector is invalid:
```
                let path = jsonpath::parse(&t.selector);

                if let Ok(ref p) = path {
```
The if` statement then tests that parsing failed if and only if  the testcase says the selector is invalid.

**Comment by gregsdennis** [comment]
Shows how much I know about rust.

**Comment by glyn** [comment]
I think this is ready to merge now the spec is updated. @mkmik or @gregsdennis: please approve.

**Review by mkmik — APPROVED** [review]

**Pull request merged**
